pkg/sql/catalog/table_elements.go

```diff
@@ -408,30 +433,153 @@ type Constraint interface {
 	GetConstraintValidity() descpb.ConstraintValidity
 
+	// IsEnforced returns true iff the constraint is enforced for all writes to
+	// the parent table.
+	IsEnforced() bool
+
 	// GetName returns the name of this constraint update mutation.
 	GetName() string
 
-	// NotNullColumnID returns the underlying not-null column ID, if there is one.
-	NotNullColumnID() descpb.ColumnID
+	// IsConstraintValidated returns true iff the constraint is enforced for
+	// all writes to the table data and has also been validated on the table's
+	// existing data prior to the addition of the constraint, if there was any.
+	IsConstraintValidated() bool
+
+	// IsConstraintUnvalidated returns true iff the constraint is enforced for
+	// all writes to the table data but which is explicitly NOT to be validated
+	// on any data in the table prior to the addition of the constraint.
+	IsConstraintUnvalidated() bool
+}
+
+// UniqueConstraint is an interface for a unique constraint.
+type UniqueConstraint interface {
+	Constraint
+
+	// IsValidReferencedUniqueConstraint returns whether the unique constraint can
+	// serve as a referenced unique constraint for a foreign key constraint.
+	IsValidReferencedUniqueConstraint(fk ForeignKeyConstraint) bool
+
+	// NumKeyColumns returns the number of columns in this unique constraint.
+	NumKeyColumns() int
+
+	// GetKeyColumnID returns the ID of the column in the unique constraint at
+	// ordinal `columnOrdinal`.
+	GetKeyColumnID(columnOrdinal int) descpb.ColumnID
+
+	// CollectKeyColumnIDs returns the columns in the unique constraint in a new
+	// TableColSet.
+	CollectKeyColumnIDs() TableColSet
+
+	// IsPartial returns true iff this is a partial index.
+	IsPartial() bool
+
+	// GetPredicate returns the partial predicate if there is one, "" otherwise.
+	GetPredicate() string
+}
+
+// UniqueWithIndexConstraint is an interface around a unique constraint
+// which backed by an index.
+type UniqueWithIndexConstraint interface {
+	UniqueConstraint
+	Index
+}
+
+// WithoutIndexConstraint is the supertype of all constraint subtypes which are
+// not backed by an index.
+type WithoutIndexConstraint interface {
+	Constraint
+}
+
+// CheckConstraint is an interface around a check constraint.
+type CheckConstraint interface {
+	WithoutIndexConstraint
+
+	// CheckDesc returns the underlying descriptor protobuf.
+	CheckDesc() *descpb.TableDescriptor_CheckConstraint
+
+	// GetExpr returns the check expression as a string.
+	GetExpr() string
+
+	// NumReferencedColumns returns the number of columns referenced in the check
+	// expression.
+	NumReferencedColumns() int
+
+	// GetReferencedColumnID returns the ID of the column referenced in the check
+	// expression at ordinal `columnOrdinal`.
+	GetReferencedColumnID(columnOrdinal int) descpb.ColumnID
+
+	// CollectReferencedColumnIDs returns the columns referenced in the check
+	// constraint expression in a new TableColSet.
+	CollectReferencedColumnIDs() TableColSet
```

Comment:
Unlike all other constraints, it's perfectly fine to have duplicate columns in a CHECK constraint's expression. Do those interface methods return the un-deduplicated column list in its implementation? I think we should, and it'd be nice if we also have some comments here explaining that.